### PR TITLE
Ragged batching interface for SonicTriton

### PR DESCRIPTION
**Pull request by kpedro88**
#### PR description:

Triton finally added a client-side friendly interface for ragged batching (processing requests with different shapes together). This PR adapts SonicTriton internally to switch between rectangular batching and ragged batching automatically, depending on how shapes are specified, without the user (module developer) having to worry about it. An option to switch manually is also provided. The changes are backwards-compatible: new functionality is available, but existing code that does not use the new functionality will continue to work as before. The README is updated to explain these different cases and the new features. The default Triton server image is also moved to a newer version. This PR may resolve several of the occasional IB failures (etc.) that have been noted in past months; these will be rechecked once it is merged.

Technical details: in Triton, a "request" consists of data that all have the same shape. Therefore, to send multiple inputs with different shapes at the same time, multiple requests must be created. Triton now provides `InferMulti()` and `AsyncInferMulti()` interfaces that take a vector of requests and loop over them, in order to handle this more general case.

Contingent details: this interface was actually implemented in Summer 2022, specifically intended to be used with ParticleNet (where we currently pad all inputs to achieve a uniform shape). However, because of various limitations in both PyTorch and ONNX, the performance in the ParticleNet case currently does not improve when using ragged batching. We expect this will eventually be resolved. In the meantime, we want to merge this updated interface so further development can continue on top of these rather involved changes.

#### PR validation:

* SonicTriton unit tests pass (including new unit test specifically for ragged batching functionality)
* DRN unit test passes
* DRN matrix workflow 10805.31 passes

### CODE REVIEW

**Comment by kpedro88** [comment]
test parameters
* pull_requests = https://github.com/cms-sw/cmsdist/pull/8324, https://github.com/cms-data/HeterogeneousCore-SonicTriton/pull/2, https://github.com/cms-data/RecoEgamma-EgammaPhotonProducers/pull/2, https://github.com/cms-data/RecoEcal-EgammaClusterProducers/pull/3
* workflows = 10805.31

**Comment by kpedro88** [comment]
@smuzaffar I think this is encountering the issue where code-checks won't proceed correctly until the necessary external is rebuilt... remind me how we fix that?

**Comment by smuzaffar** [comment]
@kpedro88 , is such cases you need to first run the PR tests for cmsdist. Once cmsdist PR has built cmssw-tool-conf then you can comment `code-checks with cms.weekN.PR_hash/tool-conf-version`  ( note that cmssw-tool-conf hash/version should be available via cmsdist PR summary page). 

I have started the tests for https://github.com/cms-sw/cmsdist/pull/8324.

**Comment by smuzaffar** [comment]
code-checks with cms.week0.PR_90a96665/53.0-0c1373420203c71a39be3b3fd9d418aa

**Comment by kpedro88** [comment]
please test

**Comment by kpedro88** [comment]
The comparison changes come from the inclusion of #40644, not from this PR.

File: HeterogeneousCore/SonicTriton/src/TritonClient.cc [hunk 1]
```diff
@@ -38,14 +38,16 @@ namespace {
...
 TritonClient::TritonClient(const edm::ParameterSet& params, const std::string& debugName)
     : SonicClient(params, debugName, "TritonClient"),
+      batchMode_(TritonBatchMode::Rectangular),
+      manualBatchMode_(false),
       verbose_(params.getUntrackedParameter<bool>("verbose")),
       useSharedMemory_(params.getUntrackedParameter<bool>("useSharedMemory")),
-      compressionAlgo_(getCompressionAlgo(params.getUntrackedParameter<std::string>("compression"))),
-      options_(params.getParameter<std::string>("modelName")) {
+      compressionAlgo_(getCompressionAlgo(params.getUntrackedParameter<std::string>("compression"))) {
+  options_.emplace_back(params.getParameter<std::string>("modelName"));
```
Comment (makortel) [inline HeterogeneousCore/SonicTriton/src/TritonClient.cc:46]: Would
```suggestion
      options_{params.getParameter<std::string>("modelName"},
      compressionAlgo_(getCompressionAlgo(params.getUntrackedParameter<std::string>("compression"))) {
```
work?

File: HeterogeneousCore/SonicTriton/src/TritonClient.cc [hunk 2]
```diff
@@ -306,10 +385,10 @@ void TritonClient::evaluate() {
       return;
   } else {
     //blocking call
-    tc::InferResult* results;
+    std::vector<tc::InferResult*> results;
     success = handle_exception([&]() {
       TRITON_THROW_IF_ERROR(
-          client_->Infer(&results, options_, inputsTriton_, outputsTriton_, headers_, compressionAlgo_),
+          client_->InferMulti(&results, options_, inputsTriton, outputsTriton, headers_, compressionAlgo_),
           "evaluate(): unable to run and/or get result");
```
Comment (makortel) [inline HeterogeneousCore/SonicTriton/src/TritonClient.cc:392]: Based on the code on `getResults()` I guess `results` is the owner of the pointers it contains. How about converting the vector of raw pointers to vector of smart pointers already here to make make their ownership clear.

File: HeterogeneousCore/SonicTriton/src/TritonClient.cc [hunk 3]
```diff
@@ -222,30 +267,63 @@ bool TritonClient::handle_exception(F&& call) {
...
-    if (output.variableDims()) {
-      std::vector<int64_t> tmp_shape;
-      TRITON_THROW_IF_ERROR(results->Shape(oname, &tmp_shape), "getResults(): unable to get output shape for " + oname);
-      if (!noBatch_)
-        tmp_shape.erase(tmp_shape.begin());
-      output.setShape(tmp_shape);
-      output.computeSizes();
+void TritonClient::getResults(std::vector<tc::InferResult*>& results) {
+  for (unsigned i = 0; i < results.size(); ++i) {
+    std::shared_ptr<tc::InferResult> result(results[i]);
```
Comment (makortel) [inline HeterogeneousCore/SonicTriton/src/TritonClient.cc:272]: Would `unique_ptr` work, or is shared ownership needed?

File: HeterogeneousCore/SonicTriton/src/TritonClient.cc [hunk 4]
```diff
@@ -325,8 +404,7 @@ void TritonClient::evaluate() {
       reportServerSideStats(stats);
     }
 
-    std::shared_ptr<tc::InferResult> results_ptr(results);
-    success = handle_exception([&]() { getResults(results_ptr); });
+    success = handle_exception([&]() { getResults(results); });
```
Comment (makortel) [inline HeterogeneousCore/SonicTriton/src/TritonClient.cc:407]: Would
```suggestion
    success = handle_exception([&]() { getResults(std::move(results)); });
```
work?

File: HeterogeneousCore/SonicTriton/src/TritonClient.cc (hunk 2 above)
Comment (kpedro88) [reply HeterogeneousCore/SonicTriton/src/TritonClient.cc:392]: The interface that requires `std::vector<InferResult*>* results` is set by Triton: https://github.com/triton-inference-server/client/blob/r22.08/src/c%2B%2B/library/grpc_client.cc#L1132

File: HeterogeneousCore/SonicTriton/src/TritonClient.cc (hunk 3 above)
Comment (kpedro88) [reply HeterogeneousCore/SonicTriton/src/TritonClient.cc:272]: Shared ownership is needed. Multiple `TritonDataEntry` objects can use the same `InferResult` object.

File: HeterogeneousCore/SonicTriton/src/TritonClient.cc (hunk 4 above)
Comment (kpedro88) [reply HeterogeneousCore/SonicTriton/src/TritonClient.cc:407]: What is the benefit of moving an `std::vector` in this case?

File: HeterogeneousCore/SonicTriton/src/TritonClient.cc (hunk 1 above)
Comment (kpedro88) [reply HeterogeneousCore/SonicTriton/src/TritonClient.cc:46]: Unfortunately the constructor is declared as explicit: https://github.com/triton-inference-server/client/blob/r22.08/src/c%2B%2B/library/common.h#L160
So trying this gives a compiler error:
```
error: converting to 'triton::client::InferOptions' from initializer list would use explicit constructor 'triton::client::InferOptions::InferOptions(const string&)'
```

The only way I can get it to work is:
```
options_{tc::InferOptions(params.getParameter<std::string>("modelName"))}
```

Is that still preferable?

File: HeterogeneousCore/SonicTriton/src/TritonClient.cc (hunk 2 above)
Comment (makortel) [reply HeterogeneousCore/SonicTriton/src/TritonClient.cc:392]: I understand the `InferMulti()` needs the vector of raw pointers, but how about converting it to vector of smart pointers right after the call returns?

File: HeterogeneousCore/SonicTriton/src/TritonClient.cc (hunk 1 above)
Comment (makortel) [reply HeterogeneousCore/SonicTriton/src/TritonClient.cc:46]: Maybe the `emplace_back()` is more clear in this case.

File: HeterogeneousCore/SonicTriton/src/TritonClient.cc (hunk 3 above)
Comment (makortel) [reply HeterogeneousCore/SonicTriton/src/TritonClient.cc:272]: Ok (right, the shared ownership is also visible in the for loop below)

File: HeterogeneousCore/SonicTriton/src/TritonClient.cc [hunk 5]
```diff
@@ -306,10 +385,10 @@ void TritonClient::evaluate() {
...
     //blocking call
-    tc::InferResult* results;
+    std::vector<tc::InferResult*> results;
     success = handle_exception([&]() {
       TRITON_THROW_IF_ERROR(
-          client_->Infer(&results, options_, inputsTriton_, outputsTriton_, headers_, compressionAlgo_),
+          client_->InferMulti(&results, options_, inputsTriton, outputsTriton, headers_, compressionAlgo_),
           "evaluate(): unable to run and/or get result");
     });
     if (!success)
```
[the remark this answers is not in the document]
Comment (kpedro88) [reply HeterogeneousCore/SonicTriton/src/TritonClient.cc:407]: Ah, I understand your point now. I've made the several related updates to ensure that memory is managed appropriately everywhere.

**Comment by kpedro88** [comment]
code-checks with cms.week1.PR_bac95210/54.0-e398b287c3659bd830ac1c4623280ff0

**Comment by kpedro88** [comment]
please test

**Comment by makortel** [comment]
+heterogeneous

**Comment by rappoccio** [comment]
+1

**Comment by makortel** [comment]
@rappoccio I suppose the dependent external and data PRs need to be merged too.

**Comment by kpedro88** [comment]
Yes, all the PRs listed in https://github.com/cms-sw/cmssw/pull/40814#issuecomment-1435140346 need to be merged together or the next IB will fail.

**Comment by smuzaffar** [comment]
@rappoccio @perrotta , this PR requires the following PRs too
https://github.com/cms-sw/cmsdist/pull/8324
https://github.com/cms-data/HeterogeneousCore-SonicTriton/pull/2
https://github.com/cms-data/RecoEgamma-EgammaPhotonProducers/pull/2
https://github.com/cms-data/RecoEcal-EgammaClusterProducers/pull/3

**Comment by perrotta** [comment]
@kpedro88 tests show a couple of modified user-floats changes for slimmed pat photons. Does it have anything to do with the modifications implemented with https://github.com/cms-data/RecoEgamma-EgammaPhotonProducers/pull/2, which was [supposed](https://github.com/cms-data/RecoEgamma-EgammaPhotonProducers/pull/2#issue-1574495749) to be  only a "minor syntactic change"?

**Comment by kpedro88** [comment]
@perrotta thanks for pointing these out, I had missed them. It may be due to the data update. This special workflow 10805.31 isn't yet run in production. @ssrothman can you take a look?

**Comment by perrotta** [comment]
Thank you @kpedro88.
Maintaining this PR without the additional external data will break the IBs (it is already breaking some tests indeed). Would you suggest to merge the externals anyhow, or revert this one till the end of your investigations and re-revert it afterward? I would rather revert it and merge only when we are sure about its outputs (so that possible mistakes do not remain unattended): but please let us know if you have a better plan, or if you think that your investigations can be quick enough,

**Comment by kpedro88** [comment]
@perrotta let's merge the externals now if that's okay. These will just go into the first prerelease, and we will definitely fix any unintended regressions before 13_1_X is final. Other development needs to start on top of the code and external changes here.

**Comment by perrotta** [comment]
> @perrotta let's merge the externals now if that's okay. These will just go into the first prerelease, and we will definitely fix any unintended regressions before 13_1_X is final. Other development needs to start on top of the code and external changes here.

Uhm, wouldn't it be easier to identify such a possible regression if there is a "clean" baseline to compare with?

**Comment by kpedro88** [comment]
For two photon userfloats, it's easy enough to run tests in an older release and compare by hand. (And actually I see that pre1 is already out, so this can be done with pre1 vs. pre2.)

**Comment by perrotta** [comment]
> For two photon userfloats, it's easy enough to run tests in an older release and compare by hand. (And actually I see that pre1 is already out, so this can be done with pre1 vs. pre2.)

Well, those two photon userfloats could point to some deeper unintended regression which is not limited to them...

I think it is better to have this reverted. A revert of the revert can be submitted at the same time, and it can be merged as soon as we are satisfied with the checks. We still have some time before pre2 for it.

**Comment by kpedro88** [comment]
If you insist on reverting, it's ultimately your prerogative. But I have seen these reversion/unreversion cycles introduce their own confusion and regressions. Historically, our policy was that regressions in prereleases were allowed in order to facilitate ongoing development.

**Comment by perrotta** [comment]
@kpedro88 we discussed it with Sal. Even if not optimal, let have this merged for now, together with all needed externals.
We're going to open a github issue at the same time, to be cleared by pre2 at most, No further development will be allowed in that packege till then.

**Comment by rappoccio** [comment]
@kpedro88 we can do the merge, but will make an issue marked as "urgent" #40938. Can you please fix it ASAP (before the next pre release) so we can ensure that there is nothing awry? 

Also, @cms-sw/reconstruction-l2, we will bypass signatures on 
https://github.com/cms-data/RecoEcal-EgammaClusterProducers/pull/3
https://github.com/cms-data/RecoEgamma-EgammaPhotonProducers/pull/2

but assign to take a look in the issue.

**Comment by kpedro88** [comment]
many thanks @perrotta @rappoccio ! we are actively working on fixing any regressions and should have another PR very soon.

**Pull request merged**